financial-templates-lib/price-feed/BalancerPriceFeed.js

```diff
@@ -0,0 +1,36 @@
+const assert = require("assert");
+const { PriceFeedInterface } = require("./PriceFeedInterface");
+
+class BalancerPriceFeed extends PriceFeedInterface {
+  constructor(logger, web3, getTime, abi, address, tokenIn, tokenOut) {
+    assert(tokenIn, "BalancerPriceFeed requires tokenIn");
+    assert(tokenOut, "BalancerPriceFeed requires tokenOut");
+    super();
+    this.logger = logger;
+    this.web3 = web3;
+    this.getTime = getTime;
+
+    this.contract = new web3.eth.Contract(abi, address);
+    this.currentPrice = null;
+    this.lastUpdateTime = null;
+    this.tokenIn = tokenIn;
+    this.tokenOut = tokenOut;
+  }
+  getHistoricalPrice(time) {
```

Comment:
I know you want to leave this for a separate PR but just wanted to discuss ways to implement this. Could it be implemented by simply calling the `getSpotPriceSansFee` at a given block number? or alternatively, would you want to use events in the same way that the `UniswapPriceFeed` does by looking at past instances of `LOG_SWAP` from the [Balancer contract](https://github.com/balancer-labs/balancer-core/blob/f4ed5d65362a8d6cec21662fb6eae233b0babc1f/contracts/BPool.sol#L28)?

This functionality is pretty critical for most uses of this price feed. For example, the Synthetic Peg monitor will require [this](https://github.com/UMAprotocol/protocol/blob/0fe67182ecc811c8981b901856b1ca5ff6dc6f63/monitors/SyntheticPegMonitor.js#L254) to correctly check for price feed volatility over a given window.

Reply:
Yeah, this is tricky because you can only do block number queries up to 30 minutes in the past without an archival node. Even if we were to enable archival data on Infura for our own usage, it would make usage of this price feed impractical for someone running on the infura free-tier (which is something we'd _like_ to enable).

Does balancer publish events?

Reply:
it seems like the price history thing is a big issue so i can consider adding that to this PR. I will need to look into it some more.

Reply:
Referencing [this comment](https://github.com/UMAprotocol/protocol/pull/1792#discussion_r464518201) (which I think was intended for this thread). There are a few approaches (in no particular order):
1. Use a subgraph (not sure how easy it is to access price history there)?
2. Get the current ratios of tokenA to tokenB, then walk backwards looking at token Transfer() events to and from the Balancer pool address on each of those tokens until you get to the historical point you want. That will give you the pool ratio of tokenA to tokenB at the time you care about -- then you can denormalize by the balancer pool weighting to figure out the price. Like you mentioned, this state reconstruction is expensive and complicated.
3. Find the SWAP event that immediately preceded the time point you care about and take the average price of that swap. This will be somewhat inaccurate since the average price of the swap is not the prevailing price that's left _after_ the swap, but it might be good enough™️. You may also need to take fees into account.
4. Do a historical query of `getSpotPriceSansFee` as @chrismaree suggested, but gracefully handle when a price that far back isn't available.

Thoughts?

Reply:
1. i looked into subgraph originally and i concluded that i couldnt get the historical price directly. But it is possible I think to get the events that would lead up to the price. Similar to using web3 logs. 
2. i think its easier to reason about events as a fast forward than a rewind, i think too complicated. 
3. swap event might be easily queried through subgraph, if so this is viable.
4. im not familiar with this feature. is this similar to what nick did with the bots state issue, by specifying a block number? If so maybe this would require the least code.

I think solution 4 is probably the best with maybe a fallback to 3. 

My biggest hesitation with this state reconstruction is that it seems wrong to duplicate code in js that already exists in the balancer solidity contract. it also introduces a new category of potential bugs, that is off chain logic is subtly different than on chain logic, or that its really hard to verify correctness. 

Id feel better if we could do one of these instead:
1. we could somehow leverage the actual balancer code to recreate state. 
2. we could snapshot historical data each block from the actual contract and persist it.
3. send balancer subgraph a PR to add block queries to their api
4. see if this is a common problem and solve it holistically rather than a 1 off solution. Abstract solution to include any type of historical data we may need from contracts.

Reply:
> 3. swap event might be easily queried through subgraph, if so this is viable.

We could also do this directly through the event queries, right? Or are you saying that is too expensive/hard to reason about?

> 4. im not familiar with this feature. is this similar to what nick did with the bots state issue, by specifying a block number?

Yep, exactly.

> I think solution 4 is probably the best with maybe a fallback to 3.

Yeah, I think that makes sense.

> My biggest hesitation with this state reconstruction is that it seems wrong to duplicate code in js that already exists in the balancer solidity contract. it also introduces a new category of potential bugs, that is off chain logic is subtly different than on chain logic, or that its really hard to verify correctness.

100% agree. This has been problematic for us in the past.

> 4. see if this is a common problem and solve it holistically rather than a 1 off solution. Abstract solution to include any type of historical data we may need from contracts.

The out-of-the-box solutions that I'm aware of are to use/run an archival node or TheGraph, which both have their tradeoffs (interested to hear if there are other out-of-the-box solutions for this, though!). Arbitrary historical contract state could be stored on a case-by-case basis using an in-house solution, too, which would use fewer resources, but would need to be built, indexed, and persisted for us and others to query for it to really make these bots usable for anyone. Would be interested to hear what you think the solution space looks like for this holistic approach.

Reply:
> We could also do this directly through the event queries, right? Or are you saying that is too expensive/hard to reason about?
Yes you could also do this with logs! I think its better to do it with logs to avoid adding subgraph dependency.

> Would be interested to hear what you think the solution space looks like for this holistic approach.
I'd have to think on this some more and identify the constraints and assumptions we have for our use case which might simplify the solution. I think though the only solution that requires no persistence (outside blockchain) is state reconstruction from logs unfortunately. So we have to choose between that or some SaaS solution (roll our own snapshots, balancer subgraph, archive node). None of which are great, but I would probably lean towards a service which snapshots state for us, probably a custom subgraph or node service + db. 

Zooming back into this particular problem, I think if we are ok with `good enough` then we go with what you suggested originally. Lean on infura for history and then fall back to logs which may be approximate. Kick the can down the road for a more robust solution. 

What do u think?

Reply:
I think your reasoning is spot-on, here. Agree with everything above.

I think leaning on infura for history works for now. I don't think we need to necessarily build an event fallback unless you think that's necessary. I would argue just returning `null` if the query is too far back is probably good enough for now, other price feeds in the repo have that behavior if you request a price outside their range.